internal/dispatch/graph/computecheck_test.go

```diff
@@ -610,9 +618,10 @@ func TestComputeCheckWithCaveats(t *testing.T) {
 			definition user {}`,
 			map[string]caveatDefinition{
 				"expired": {
-					"now() < timestamp(expiration)",
```

Comment:
i added the `now()` function as I thought it was better UX than having to provide it as part of the request context. Why removing it?

Reply:
@josephschorr I may be missing something here, but this seems contradictory to me.

There is nothing stateless about caveats: state is provided by the request, and state is stored in the relationship, and even the relationship itself is state. If state can cause problems with caching, how is it different from providing `now` in the request context, from having `now` stored as relationship context, from invoking a function?

The only challenge I see here is that the value of `now()` changes depending on where you invoke it, because different executions of `now()` within a request can yield different results, providing unpredictable results. We can address this making sure `now()` returns the same value across all executions in the same request, and that would make it not that much different from providing it in the request context. This obviously raises the question that a request (e.g. `LookupResources`) could take enough time that `now()` being constant could be inaccurate_, but it seems like the safest approach.

Effectively we can turn `now()` into `request_timestamp()` - it may be easier to understand for SpiceDB users. I don't see any practical difference with providing that in the request context. And I'd argue, in fact, that it would be desirable to have the server compute this value, given clocks between client and server could be vastly different.

Reply:
> @josephschorr I may be missing something here, but this seems contradictory to me.
> 
> There is nothing stateless about caveats: state is provided by the request, and state is stored in the relationship, and even the relationship itself is state. If state can cause problems with caching, how is it different from providing `now` in the request context, from having `now` stored as relationship context, from invoking a function?

All the state of those caveats is *client provided*, which means the client has full control over the execution context at all times. Adding server-side state breaks that.

> The only challenge I see here is that the value of `now()` changes depending on where you invoke it, because different executions of `now()` within a request can yield different results, providing unpredictable results. We can address this making sure `now()` returns the same value across all executions in the same request, and that would make it not that much different from providing it in the request context. This obviously raises the question that a request (e.g. `LookupResources`) could take enough time that `now()` being constant could be inaccurate_, but it seems like the safest approach.

Won't work if we move some of the evaluation into the dispatch, which may happen down the road for theoretical performance reasons; caching would result in the value being stored, which would break it. It *has* to either be provided by the API layer or has to be provided by the caller, and caller is cleaner for reasons I've outlined below.

> 
> Effectively we can turn `now()` into `request_timestamp()` - it may be easier to understand for SpiceDB users. I don't see any practical difference with providing that in the request context. 

It has two major differences: repeatability and control. If I wish to issue a Check with the *exact same* context as a previous check, I can do so by re-specifying `now` in my context to that I previously gave, and it'll Just Work (TM). This is *very* important for issuing a Check to narrow a previous Check, or to issue a Check on a previously caveated result, or for Checking after getting back a "possible" on a LookupResources or for pagination of results, etc. However, a server-side provided value is, by definition, not controllable and not repeatable: the caller has no control over the value. This also makes debugging and validation/testing MUCH harder, if not impossible, without special handling: you can't easily just test with a different time. For example, how would you write a validation test with a caveated relationship based on it being a specific time, if the `now()` is computed at the time the test runs?

> And I'd argue, in fact, that it would be desirable to have the server compute this value, given clocks between client and server could be vastly different.

I'd argue that is a major reason to *not* do so server-side: we provide no guarantees on the clocks for SpiceDB itself and if the caller needs high level precision on its clocks, then that's an exercise left for them. If, down the road, super high precision clocks become an important caveat feature, then I'd imagine we'd define a function at the API level, but it would *still* have to be overridable from the caller for the purposes of debugging and testing

Reply:
Thanks for the detailed answer @josephschorr, these are solid arguments and you convinced me 👍🏻